allennlp/models/semantic_parsing/text2sql_parser.py

```diff
@@ -0,0 +1,399 @@
+            The output of ``TextField.as_array()`` applied on the tokens ``TextField``. This will
+            be passed through a ``TextFieldEmbedder`` and then through an encoder.
+        valid_actions : ``List[List[ProductionRule]]``
+            A list of all possible actions for each ``World`` in the batch, indexed into a
+            ``ProductionRule`` using a ``ProductionRuleField``.  We will embed all of these
+            and use the embeddings to determine which action to take at each timestep in the
+            decoder.
+        target_action_sequence : torch.Tensor, optional (default=None)
+            The action sequence for the correct action sequence, where each action is an index into the list
+            of possible actions.  This tensor has shape ``(batch_size, sequence_length, 1)``. We remove the
+            trailing dimension.
+        sql_queries : List[List[str]], optional (default=None)
+            A list of the SQL queries that are given during training or validation.
+        """
+        embedded_utterance = self._utterance_embedder(tokens)
+        mask = util.get_text_field_mask(tokens).float()
+        batch_size = embedded_utterance.size(0)
+
+        # (batch_size, num_tokens, encoder_output_dim)
+        encoder_outputs = self._dropout(self._encoder(embedded_utterance, mask))
+        initial_state = self._get_initial_state(encoder_outputs, mask, valid_actions)
+
+        if action_sequence is not None:
+            # Remove the trailing dimension (from ListField[ListField[IndexField]]).
+            action_sequence = action_sequence.squeeze(-1)
+            target_mask = action_sequence != self._action_padding_index
+        else:
+            target_mask = None
+
+        outputs: Dict[str, Any] = {}
+        if action_sequence is not None:
+            # target_action_sequence is of shape (batch_size, 1, target_sequence_length)
+            # here after we unsqueeze it for the MML trainer.
+            loss_output = self._decoder_trainer.decode(initial_state,
+                                                       self._transition_function,
+                                                       (action_sequence.unsqueeze(1),
+                                                        target_mask.unsqueeze(1)))
+            outputs.update(loss_output)
+
+        if not self.training:
+            action_mapping = []
+            for batch_actions in valid_actions:
+                batch_action_mapping = {}
+                for action_index, action in enumerate(batch_actions):
+                    batch_action_mapping[action_index] = action[0]
+                action_mapping.append(batch_action_mapping)
+
+            outputs['action_mapping'] = action_mapping
+            # This tells the state to start keeping track of debug info, which we'll pass along in
+            # our output dictionary.
+            initial_state.debug_info = [[] for _ in range(batch_size)]
```

Comment:
Looking at this, we should probably just have a method to call on the `State` object - would be easier to understand.  Not for this PR necessarily, just a note to myself.